Adding PR Guidelines to workflow (#6307)
NuGet/NuGet.Client: docs/workflow.md

### Merging Pull Requests
- **Required**: All comments must be resolved before merging.
- **Required**: use the GitHub `Squash and Merge` button for this repository.
- Merge a PR when you are confident it has been sufficiently reviewed.
- When possible wait 24h after the last significant change before merging.
- Ensure commit message is descriptive and helpful
![Good Commit Message](images/good-commit-message.png)
![Good Commit Message With More Details](images/good-commit-message-expanded.png)
![Bad Commit Message](images/bad-commit-message.png)

### Providing Pull Request Feedback
- **Required**: Focus on collaboration when providing feedback.
- Use [suggested changes](https://docs.github.com/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request), but please never commit them for a PR you did not create.
- Make the intent of your comment clear.
  - ex. "Consider changing the color of this boarder because..."
  - ex. "I've never seen this before! Can you help me understand..."
  - ex. "Nit: Spelling mistake"
- Focus on actionable feedback when requesting changes.

>   - ❌ This button won't do anything

Reviewer: I see this as being actionable. Requiring a reviewer to do the investigation and provide exact steps to remediate seems unreasonable. Perhaps a better example can be added?

Author: I would say that it isn't actionable feedback. While the author might now know the button isn't working, the reviewer hasn't shared the context as to _why_ it wouldn't work. For this example, I don't think there's really any additional investigation if a button is missing a listener.

Reviewer 2: In more general terms, comments such as "this code won't work" are not very actionable code review comments. 

According to Copilot:

The comment "this button won't do anything" is not very actionable because it doesn't provide specific details about why the button won't work or what needs to be done to fix it. An actionable code review comment should offer clear, constructive feedback that helps the developer understand the issue and how to resolve it.

An alternative, more actionable comment could be:

"This button won't do anything because the click event handler is not defined. You need to add an event listener for the button's click event that triggers the desired action."

This comment is actionable because it:

Identifies the specific issue (missing click event handler).
Provides a suggestion for how to fix the issue (adding an event listener).
Would you like more examples of actionable code review comments or any other assistance with code reviews?

Reviewer: If someone submits a PR with a button not working, I don't expect we should have to go figure out what they didn't do to make the button work.

> While the author might now know the button isn't working

This tells me the PR author did not do basic UI testing on their work before opening the PR. Therefore, adding this as PR guidance is not an acceptable standard to me. 

Since you're also tracking metrics on review time, note that such practice will also be increasing time-to-complete PRs.

Author: I pulled up some examples from ChatGPT:

Unactionable Feedback: ❌ "Your code is messy."
Actionable Feedback: ✔️ "Consider organizing your functions into smaller, more modular components to improve readability and maintainability."

Unactionable Feedback:
❌ "You’re not using best practices here."
Actionable Feedback:
✔️ "Consider using const or let instead of var for variable declarations to avoid potential scoping issues. This can help make your code more predictable."

Unactionable Feedback:
❌ "The formatting is wrong."
Actionable Feedback:
✔️ "It seems like the indentation is inconsistent. Following the style guide (e.g., using 2 spaces for indentation) could help make the code more consistent and easier to maintain."

That said this is meant to be an example, do we want to hold examples to this level of scrutiny? Is there a concern that someone will reference this example in their PR when we ask if they tested their code? It just seems like a lot of weight to place on an example.

Reviewer 3: I feel like the some of the examples in that list are really wordy. You can be nice without writing fluff. 

That being said we're not adding those examples, but a different one. 
I think it makes sense that the examples here are something universally agreed upon. 
When using strong wording in guidelines like do this and don't do that, people are likely to take it very seriously. 

I think it's worth coming up with a different example than the current one, one that doesn't put heavy burden on the reivewer.

Author: Removed the example and asked chat gpt to give me some bullet points on providing actionable feedback.

  - ✔️ You need to add a listener to the button for it to respond
- Ensure the subject of your feedback is the code and not the PR author.
- Use soft language if you must address the author directly.

#### When to mark a PR as "Approved"
- You feel confident that the code meets a high-quality bar, has adequate test coverage, and is ready to merge.
- Any comments submitted are uncontroversial and do not need to be resolved before merging.
- Previous feedback has been sufficiently addressed.
#### When to leave comments without approval
- You are not confident you can approve the PR with the requirements stated above.
- You have feedback that you would like addressed but do not want to block the PR from merging.

> #### When to mark a PR as "Request Changes"

Reviewer 4: if you edit this markdown in VS Code, and install the "markdownlint" extension, you'll get a bunch of suggestions like to add an empty line before and after any heading

# Workflow Guidelines

In here we describe the general workflow guidelines the NuGet developer/contributor.

## Pull Requests & Code Reviews

To help ensure that only the highest quality code makes its way into the project, all code changes need to be submitted to GitHub as PRs.
This repo has bots that manage all stale PRs.
Stale PRs will be auto closed.

> The following guidelines should be viewed as suggestions, unless explicitly stated they are required.

Reviewer 3: > unless explicitly stated they are required

Required label is only added in the addressing feedback part. I think there's some more requirements as well, such as getting the SME to review, or resolving all comments, things that are enforced through tooling.

Author: I can add more required tags! I was more so trying to make clear where something can be viewed as a suggestion vs a hard and fast rule. 

How do you feel about this suggestion? 

```suggestion
These are guidelines to follow but unless indicated, they are not required. 
 [❗] = Required
 [🤖] = Enforced with tooling
```

Reviewer 3: Sure, feel free to apply it. 
It literally adds some color to the whole doc :D

> ### Requesting a Pull Request

Reviewer 3: Requesting a Pull Request Review?

> - Include the Subject Matter Expert (SME) for the area you are working on. If you don't know the SME, someone on the team will help you identify them.

Reviewer 3: Can you help me understand the idea behind the ordering switch? 

Following the PR template seemed like the first thing to consider when submitting a PR. 

While the previous iteration wasn't perfect, the idea of it was that it followed somewhat of a chronological order.

Author: I updated the order

- Follow the pull request template, as it helps the maintainers drive quality across the product.
- Request at least 2 reviewers, and re-request their review when feedback has been addressed.

> - Include screenshots and PM when making UX changes.

Reviewer 3: This line and the next one go into more specifics around the types of PRs and considerations based on that. 

Things like `Focus on small, iterative, changes when possible` talk about general approaches to PRs, so I feel like that should be among the first.

My suggestion would be: 

Set ground rules about what kind of PRs to create
Then talk about the reqs for reviewing 
Finally go into detail about the types of things to provide in the PR description

- Include perf traces when making performance improvements.
- Focus on small, iterative, changes when possible.

### Addressing Feedback

- **Required**: Focus on collaboration when responding to feedback.

> - Try your best not to take feedback personally, we're all working towards the same goals.

Reviewer 3: When I was reading this, line 24 felt like it was setting up the context, and then these to be guidelines around the requirement. 

Right now, they have the same indentation. 
Should we either remove indentation on line 24 or add more on 25-28?

Author: I added indentation, that way all the guidelines are bulleted.

- When conflict arises try to address it directly with the commentor offline, preferably synchronously.
- Provide an explanation if you decide not to implement a recommendation.

> - Use GitHub's tooling. Re-request review after feedback has been addressed.

Reviewer 3: We talk about the re-request review part in multiple places, maybe we can simplify it to avoid repeating the same thing. 

I feel like the longer the guidelines, the more likely we are to miss something, so if we can say something in fewer words we should do that.

Author: Hmm that makes sense, I removed the line entirely, feels covered in the first section

### Merging Pull Requests

> - **Required**: All comments must be resolved before merging.

Reviewer 3: This is enforced by tooling.

> - **Required**: use the GitHub `Squash and Merge` button for this repository.

Reviewer 3: Same here, this is enforced by tooling.

### Merging Pull Requests

- [🤖] All comments must be resolved before merging.
- [🤖] use the GitHub `Squash and Merge` button for this repository.
- Merge a PR when you are confident it has been sufficiently reviewed.
- When possible wait 24h after the last significant change before merging.
- Ensure commit message is descriptive and helpful
![Good Commit Message](images/good-commit-message.png)
![Good Commit Message With More Details](images/good-commit-message-expanded.png)
![Bad Commit Message](images/bad-commit-message.png)

### Providing Pull Request Feedback

- [❗] Focus on collaboration when providing feedback.
- Use [suggested changes](https://docs.github.com/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request), but please never commit them for a PR you did not create.
- Make the intent of your comment clear.
  - ex. "Consider changing the color of this border because..."
  - ex. "I've never seen this before! Can you help me understand..."
  - ex. "Nit: Spelling mistake"
- Focus on actionable feedback when requesting changes.

>   - **Be Constructive**: Offer suggestions for improvement rather than just pointing out what's wrong.

Reviewer 3: There's a learning element here that is more powerful when the right questions are being asked, rather than just offering the answer every time. 
I don't know if it needs to change the description of "Be constructive", but I feel there should be room for this in PR reviews.

Author: Fair, added a line about guiding the author, thoughts?

Reviewer 3: The new version reads so well!

  - **Provide Context**: Explain why a change is necessary or how it aligns with project goals
  - **Be Specific**: Clearly point out the exact lines or sections of code that need attention.
- Ensure the subject of your feedback is the code and not the PR author.
  - Use soft language if you must address the author directly.

#### When to mark a PR as "Approved"

- You feel confident that the code meets a high-quality bar, has adequate test coverage, and is ready to merge.

> - Any comments submitted are uncontroversial and do not need to be resolved before merging.

Reviewer 3: I think we've lost the wording of: 

> If the author pushes a new change addressing those comments, you may need to re-approve as it's required that the latest iteration of a PR is approved.

I feel like this is an important aspect of keeping the process more accessible for the reviewer. 
For example, it allows me to do things like "Fix this typo in the interface name and approve", basically saying I approve of the design but I've left a comment that I trust you'll address before merging. 

If people think this is ambiguous or not a good practice we can always remove it ofc.

Author: I added a sub bullet about having to re-approve

# Workflow Guidelines

In here we describe the general workflow guidelines the NuGet developer/contributor.

## Pull Requests & Code Reviews

To help ensure that only the highest quality code makes its way into the project, all code changes need to be submitted to GitHub as PRs.
This repo has bots that manage all stale PRs.
Stale PRs will be auto closed.

These are guidelines to follow but unless indicated, they are not required.
 [❗] = Required
 [🤖] = Enforced with tooling

### Requesting a Pull Request Review

- Focus on small, iterative, changes when possible.
- Follow the pull request template, as it helps the maintainers drive quality across the product.
- [❗] Include the Subject Matter Expert (SME) for the area you are working on. If you don't know the SME, someone on the team will help you identify them.

> - Request at least 2 reviewers, and re-request their review when feedback has been addressed.

Reviewer: Doesn't NuGet Client already get added as a reviewer? Is this a change from those practices?

Author: No, just to request at least 2 specific reviewers, it's also not required.

Reviewer 3: Should the wording be assign?

![image](https://github.com/user-attachments/assets/34c45aad-c7f1-46b6-856e-69cc1effd11f)

The request already happens to the whole team.

Author: That's fair, though the intent is really to try to get at least two people to approve right? I can move it to the Merging Pull Requests section. 

something like "Try to get at approvals from at least two different people"

Reviewer 3: Are you trying to replicate what was there before or add a new policy? 

If we're replicating what was there before, then yeah, i like the whole get approvals from 2 people. 
1 is required, 2 is recommended.

> - Include screenshots and PM when making UX changes.

Reviewer: Do we expect the community contributors to talk directly to SMEs and PMs at Microsoft?

Author: I added a line about how someone on the team will help identify the PM. Similar to the SME one. I don't think this is a change in expectation on how community contributors interact with the team. The line from SME is similar to the one that was there before.

> - Include screenshots and PM when making UX changes.

Reviewer: I created a separate UI change guideline document. Other areas probably also have niche considerations for their PRs. Could this just be kept general, like "Summarize your changes, and follow relevant area guidelines for screenshots, testing, etc"?

Author: I think breaking it down here is better instead of having to search across documents when creating a PR.

- Include perf traces when making performance improvements.

### Addressing Feedback

- [❗] Focus on collaboration when responding to feedback.
  - Try your best not to take feedback personally, we're all working towards the same goals.

>   - When conflict arises try to address it directly with the commentor offline, preferably synchronously.

Reviewer: How would an external contributor talk to us offline?

Author: You can create meeting with external people if it gets to that point. This is also a guideline and not required.

  - Provide an explanation if you decide not to implement a recommendation.

### Merging Pull Requests

- [🤖] All comments must be resolved before merging.
- [🤖] use the GitHub `Squash and Merge` button for this repository.
- Merge a PR when you are confident it has been sufficiently reviewed.

> - When possible wait 24h after the last significant change before merging.

Reviewer: Is this a rule or not? The line leaves a lot of room for interpretation. Also, I recall 24h is now considered very long by our metrics.

Reviewer 3: This has always been a guidance, not a requirement. 
It was as an opportunity to allow people to review changes. 

We can always change the wording, but may we scope it out of this PR that's already making quite a few changes.

Author: No, it's a guideline and not marked as required.

> # Workflow Guidelines

Reviewer 5: Testing

In here we describe the general workflow guidelines the NuGet developer/contributor.

## Pull Requests & Code Reviews

To help ensure that only the highest quality code makes its way into the project, all code changes need to be submitted to GitHub as PRs.
This repo has bots that manage all stale PRs.
Stale PRs will be auto closed.

> These are guidelines to follow but unless indicated, they are not required.
>  [❗] = Required
>  [🤖] = Enforced with tooling

Reviewer 6: nit
```suggestion
These are guidelines to follow but unless indicated, they are not required.
- [❗] = Required
- [🤖] = Enforced with tooling
```
